cli/storage_filesystem.go

```diff
@@ -59,21 +59,24 @@ func (c *storageFilesystemFlags) connect(ctx context.Context, isNew bool) (blob.
 
 	fso.FileMode = getFileModeValue(c.connectFileMode, defaultFileMode)
 	fso.DirectoryMode = getFileModeValue(c.connectDirMode, defaultDirMode)
+	fso.DirectoryShards = initialDirectoryShards(c.connectFlat, formatVersion)
 
-	if c.connectFlat {
-		fso.DirectoryShards = []int{}
-	}
+	// nolint:wrapcheck
+	return filesystem.New(ctx, &fso, isCreate)
+}
 
-	if isNew {
-		log(ctx).Debugf("creating directory for repository: %v dir mode: %v", fso.Path, fso.DirectoryMode)
+func initialDirectoryShards(flat bool, formatVersion int) []int {
+	if flat {
+		return []int{}
+	}
 
-		if err := os.MkdirAll(fso.Path, fso.DirectoryMode); err != nil {
-			log(ctx).Errorf("unable to create directory: %v", fso.Path)
-		}
+	// when creating a repository for version 1, use fixed {3,3} sharding,
+	// otherwise old client can't read it.
+	if formatVersion == 1 {
```

Comment:
I'm confused here about the format version and compatibility matrix.

Reply:
it's enforced in the compat tests. repository created with `--format-version=1` must be readable by 0.8.4.